File: src/CSharp/ImplicitFunctions/IsoContours.cs
using System;
using System.ComponentModel;
using System.Drawing;
using System.Windows.Forms;
using System.Runtime.InteropServices;
using System.Diagnostics;
using Kitware.VTK;

namespace ActiViz.Examples {
   public class Form1 : Form {
      private System.ComponentModel.IContainer components = null;
      private vtkContourFilter _ContourFilter;
      private vtkRenderWindow _RenderWindow;
         

      public Form1() {
         InitializeComponent();
         this.Size = new Size(620, 400);
      }


      private void Form1_Load(object sender, EventArgs e) {
         try {
            IsoContours();
         }
         catch(Exception ex) {
            MessageBox.Show(ex.Message, "Exception", MessageBoxButtons.OK);
         }
      }


      private void sliderContourValue_Scroll(object sender, EventArgs e) {
         TrackBar slider = sender as TrackBar;
         if(slider != null) {
            _ContourFilter.GenerateValues(1, slider.Value, slider.Value);
            _RenderWindow.Render();
         }
      }


      private void IsoContours() {
         vtkImageData data = vtkImageData.New();
         CreateData(ref data);

         // Create an isosurface
         _ContourFilter = vtkContourFilter.New();
#if VTK_MAJOR_VERSION_5
         _ContourFilter.SetInput(data);
#else
         contourFilter.SetInputData(data);
#endif
         _ContourFilter.GenerateValues(1, 10, 10); // (numContours, rangeStart, rangeEnd)

         // Map the contours to graphical primitives
         vtkPolyDataMapper contourMapper = vtkPolyDataMapper.New();
#if VTK_MAJOR_VERSION_5
         contourMapper.SetInputConnection(_ContourFilter.GetOutputPort());
#else
         contourMapper.SetInputData(contourFilter);
#endif
         // Create an actor for the contours
         vtkActor contourActor = vtkActor.New();
         contourActor.SetMapper(contourMapper);

         // Create the outline
         vtkOutlineFilter outlineFilter = vtkOutlineFilter.New();
         vtkPolyDataMapper outlineMapper = vtkPolyDataMapper.New();
#if VTK_MAJOR_VERSION_5
         outlineFilter.SetInput(data);
         outlineMapper.SetInputConnection(outlineFilter.GetOutputPort());
#else
         outlineFilter.SetInputData(data);
         outlineMapper.SetInputData(outlineFilter);
#endif
         vtkActor outlineActor = vtkActor.New();
         outlineActor.SetMapper(outlineMapper);

         // get a reference to the renderwindow of our renderWindowControl1
         _RenderWindow = renderWindowControl1.RenderWindow;
         // renderer
         vtkRenderer renderer = _RenderWindow.GetRenderers().GetFirstRenderer();
         // set background color
         renderer.SetBackground(.2, .3, .4);
         // add our actor to the renderer
         renderer.AddActor(contourActor);
         renderer.AddActor(outlineActor);
      }


      void CreateData(ref vtkImageData data) {
         data.SetExtent(-25, 25, -25, 25, 0, 0);
#if VTK_MAJOR_VERSION_5
         data.SetNumberOfScalarComponents(1);
         data.SetScalarTypeToDouble();
#else
         data.AllocateScalars(VTK_DOUBLE,1);
#endif
         int[] extent = data.GetExtent();

         for(int y = extent[2]; y <= extent[3]; y++) {
            for(int x = extent[0]; x <= extent[1]; x++) {
               IntPtr ptr = data.GetScalarPointer(x, y, 0);
               double[] pixel = new double[] { Math.Sqrt(Math.Pow(x, 2.0) + Math.Pow(y, 2.0)) };
               Marshal.Copy(pixel, 0, ptr, 1);
            }
         }

         vtkXMLImageDataWriter writer = vtkXMLImageDataWriter.New();
         writer.SetFileName(@"c:\vtk\vtkdata-5.8.0\Data\testIsoContours.vti");
#if VTK_MAJOR_VERSION_5
         writer.SetInputConnection(data.GetProducerPort());
#else
         writer.SetInputData(data);
#endif
         writer.Write();
      }



      /// <summary>
      /// Clean up any resources being used.
      /// </summary>
      /// <param name="disposing">true if managed resources should be disposed; otherwise, false.</param>
      protected override void Dispose(bool disposing) {
         if(disposing && ( components != null )) {
            components.Dispose();
         }
         base.Dispose(disposing);
      }

      #region Windows Form Designer generated code

      /// <summary>
      /// Required method for Designer support - do not modify
      /// the contents of this method with the code editor.
      /// </summary>
      private void InitializeComponent() {
         this.splitContainer1 = new System.Windows.Forms.SplitContainer();
         this.lblSliderMaxValue = new System.Windows.Forms.Label();
         this.lblSliderMinValue = new System.Windows.Forms.Label();
         this.lblSliderTitle = new System.Windows.Forms.Label();
         this.sliderContourValue = new System.Windows.Forms.TrackBar();
         this.renderWindowControl1 = new Kitware.VTK.RenderWindowControl();
         this.splitContainer1.Panel1.SuspendLayout();
         this.splitContainer1.Panel2.SuspendLayout();
         this.splitContainer1.SuspendLayout();
         ((System.ComponentModel.ISupportInitialize)(this.sliderContourValue)).BeginInit();
         this.SuspendLayout();
         // 
         // splitContainer1
         // 
         this.splitContainer1.Dock = System.Windows.Forms.DockStyle.Fill;
         this.splitContainer1.FixedPanel = System.Windows.Forms.FixedPanel.Panel1;
         this.splitContainer1.IsSplitterFixed = true;
         this.splitContainer1.Location = new System.Drawing.Point(0, 0);
         this.splitContainer1.Name = "splitContainer1";
         // 
         // splitContainer1.Panel1
         // 
         this.splitContainer1.Panel1.Controls.Add(this.lblSliderMaxValue);
         this.splitContainer1.Panel1.Controls.Add(this.lblSliderMinValue);
         this.splitContainer1.Panel1.Controls.Add(this.lblSliderTitle);
         this.splitContainer1.Panel1.Controls.Add(this.sliderContourValue);
         // 
         // splitContainer1.Panel2
         // 
         this.splitContainer1.Panel2.Controls.Add(this.renderWindowControl1);
         this.splitContainer1.Size = new System.Drawing.Size(604, 362);
         this.splitContainer1.SplitterDistance = 213;
         this.splitContainer1.SplitterWidth = 1;
         this.splitContainer1.TabIndex = 0;
         // 
         // lblSliderMaxValue
         // 
         this.lblSliderMaxValue.AutoSize = true;
         this.lblSliderMaxValue.Location = new System.Drawing.Point(176, 69);
         this.lblSliderMaxValue.Name = "lblSliderMaxValue";
         this.lblSliderMaxValue.Size = new System.Drawing.Size(19, 13);
         this.lblSliderMaxValue.TabIndex = 1;
         this.lblSliderMaxValue.Text = "30";
         // 
         // lblSliderMinValue
         // 
         this.lblSliderMinValue.AutoSize = true;
         this.lblSliderMinValue.Location = new System.Drawing.Point(21, 69);
         this.lblSliderMinValue.Name = "lblSliderMinValue";
         this.lblSliderMinValue.Size = new System.Drawing.Size(13, 13);
         this.lblSliderMinValue.TabIndex = 1;
         this.lblSliderMinValue.Text = "1";
         // 
         // lblSliderTitle
         // 
         this.lblSliderTitle.AutoSize = true;
         this.lblSliderTitle.Location = new System.Drawing.Point(18, 11);
         this.lblSliderTitle.Name = "lblSliderTitle";
         this.lblSliderTitle.Size = new System.Drawing.Size(74, 13);
         this.lblSliderTitle.TabIndex = 1;
         this.lblSliderTitle.Text = "Contour Value";
         // 
         // sliderContourValue
         // 
         this.sliderContourValue.Location = new System.Drawing.Point(12, 37);
         this.sliderContourValue.Maximum = 30;
         this.sliderContourValue.Minimum = 1;
         this.sliderContourValue.Name = "sliderContourValue";
         this.sliderContourValue.Size = new System.Drawing.Size(186, 45);
         this.sliderContourValue.TabIndex = 0;
         this.sliderContourValue.TickFrequency = 5;
         this.sliderContourValue.Value = 10;
         this.sliderContourValue.Scroll += new System.EventHandler(this.sliderContourValue_Scroll);
         // 
         // renderWindowControl1
         // 
         this.renderWindowControl1.AddTestActors = false;
         this.renderWindowControl1.Dock = System.Windows.Forms.DockStyle.Fill;
         this.renderWindowControl1.Location = new System.Drawing.Point(0, 0);
         this.renderWindowControl1.Name = "renderWindowControl1";
         this.renderWindowControl1.Size = new System.Drawing.Size(390, 362);
         this.renderWindowControl1.TabIndex = 0;
         this.renderWindowControl1.TestText = null;
         // 
         // Form1
         // 
         this.AutoScaleDimensions = new System.Drawing.SizeF(6F, 13F);
         this.AutoScaleMode = System.Windows.Forms.AutoScaleMode.Font;
         this.ClientSize = new System.Drawing.Size(604, 362);
         this.Controls.Add(this.splitContainer1);
         this.Name = "Form1";
         this.Text = "IsoContours";
         this.Load += new System.EventHandler(this.Form1_Load);
         this.splitContainer1.Panel1.ResumeLayout(false);
         this.splitContainer1.Panel1.PerformLayout();
         this.splitContainer1.Panel2.ResumeLayout(false);
         this.splitContainer1.ResumeLayout(false);
         ((System.ComponentModel.ISupportInitialize)(this.sliderContourValue)).EndInit();
         this.ResumeLayout(false);

      }

      #endregion

      private System.Windows.Forms.SplitContainer splitContainer1;
      private System.Windows.Forms.Label lblSliderTitle;
      private System.Windows.Forms.TrackBar sliderContourValue;
      private Kitware.VTK.RenderWindowControl renderWindowControl1;
      private System.Windows.Forms.Label lblSliderMaxValue;
      private System.Windows.Forms.Label lblSliderMinValue;
   }
}

File: src/Cxx/ImplicitFunctions/IsoContours.cxx
#include <vtkSmartPointer.h>
#include <vtkContourFilter.h>

#include <vtkInteractorStyleUser.h>
#include <vtkProperty.h>
#include <vtkOutlineFilter.h>
#include <vtkCommand.h>
#include <vtkSliderWidget.h>
#include <vtkSliderRepresentation.h>
#include <vtkSliderRepresentation3D.h>
#include <vtkImageData.h>
#include <vtkCellArray.h>
#include <vtkFloatArray.h>
#include <vtkPointData.h>
#include <vtkPolyDataMapper.h>
#include <vtkActor.h>
#include <vtkRenderer.h>
#include <vtkRenderWindow.h>
#include <vtkRenderWindowInteractor.h>
#include <vtkXMLImageDataWriter.h>
#include <vtkNamedColors.h>


namespace
{
void CreateData(vtkImageData* data);

class vtkSliderCallback : public vtkCommand
{
public:
  static vtkSliderCallback *New()
  {
    return new vtkSliderCallback;
  }
  virtual void Execute(vtkObject *caller, unsigned long, void*)
  {
    vtkSliderWidget *sliderWidget =
      reinterpret_cast<vtkSliderWidget*>(caller);
    double value = static_cast<vtkSliderRepresentation *>(sliderWidget->GetRepresentation())->GetValue();
    this->ContourFilter->GenerateValues(1, value, value);

  }
  vtkSliderCallback():ContourFilter(NULL) {}
  vtkContourFilter* ContourFilter;
};
}

int main(int, char *[])
{
  vtkSmartPointer<vtkNamedColors> colors =
    vtkSmartPointer<vtkNamedColors>::New();

  vtkSmartPointer<vtkImageData> data =
    vtkSmartPointer<vtkImageData>::New();
  CreateData(data);

  // Create an isosurface
  vtkSmartPointer<vtkContourFilter> contourFilter =
    vtkSmartPointer<vtkContourFilter>::New();
  contourFilter->SetInputData(data);
  contourFilter->GenerateValues(1, 10, 10); // (numContours, rangeStart, rangeEnd)

  // Map the contours to graphical primitives
  vtkSmartPointer<vtkPolyDataMapper> contourMapper =
    vtkSmartPointer<vtkPolyDataMapper>::New();
  contourMapper->SetInputConnection(contourFilter->GetOutputPort());

  // Create an actor for the contours
  vtkSmartPointer<vtkActor> contourActor =
    vtkSmartPointer<vtkActor>::New();
  contourActor->SetMapper(contourMapper);
  contourActor->GetProperty()->SetLineWidth(5);

  // Create the outline
  vtkSmartPointer<vtkOutlineFilter> outlineFilter =
    vtkSmartPointer<vtkOutlineFilter>::New();
  outlineFilter->SetInputData(data);

  vtkSmartPointer<vtkPolyDataMapper> outlineMapper =
    vtkSmartPointer<vtkPolyDataMapper>::New();
  outlineMapper->SetInputConnection(outlineFilter->GetOutputPort());
  vtkSmartPointer<vtkActor> outlineActor =
    vtkSmartPointer<vtkActor>::New();
  outlineActor->SetMapper(outlineMapper);
  outlineActor->GetProperty()->SetColor(colors->GetColor3d("Gray").GetData());
  outlineActor->GetProperty()->SetLineWidth(3);

  // Visualize
  vtkSmartPointer<vtkRenderer> renderer =
    vtkSmartPointer<vtkRenderer>::New();
  vtkSmartPointer<vtkRenderWindow> renderWindow =
    vtkSmartPointer<vtkRenderWindow>::New();
  renderWindow->AddRenderer(renderer);
  vtkSmartPointer<vtkRenderWindowInteractor> interactor =
    vtkSmartPointer<vtkRenderWindowInteractor>::New();
  interactor->SetRenderWindow(renderWindow);

  renderer->AddActor(contourActor);
  renderer->AddActor(outlineActor);
  renderer->SetBackground(colors->GetColor3d("Burlywood").GetData());

  vtkSmartPointer<vtkSliderRepresentation3D> sliderRep =
    vtkSmartPointer<vtkSliderRepresentation3D>::New();
  sliderRep->SetMinimumValue(0.0);
  sliderRep->SetMaximumValue(30.0);
  sliderRep->SetValue(10.0);
  sliderRep->SetTitleText("Contour value");
  sliderRep->SetPoint1InWorldCoordinates(-20, -40, 0);
  sliderRep->SetPoint2InWorldCoordinates(0, -40, 0);
  sliderRep->SetSliderWidth(.2);
  sliderRep->SetLabelHeight(.1);

  vtkSmartPointer<vtkSliderWidget> sliderWidget =
    vtkSmartPointer<vtkSliderWidget>::New();
  sliderWidget->SetInteractor(interactor);
  sliderWidget->SetRepresentation(sliderRep);
  sliderWidget->SetAnimationModeToAnimate();
  sliderWidget->EnabledOn();

  vtkSmartPointer<vtkSliderCallback> callback =
    vtkSmartPointer<vtkSliderCallback>::New();
  callback->ContourFilter = contourFilter;

  sliderWidget->AddObserver(vtkCommand::InteractionEvent,callback);

  vtkSmartPointer<vtkInteractorStyleUser> style =
    vtkSmartPointer<vtkInteractorStyleUser>::New();
  interactor->SetInteractorStyle(style);

  renderWindow->SetSize(500,500);
  renderWindow->Render();
  renderWindow->Render();
  interactor->Start();

  return EXIT_SUCCESS;
}
namespace
{
void CreateData(vtkImageData* data)
{
  data->SetExtent(-25,25,-25,25,0,0);
  data->AllocateScalars(VTK_DOUBLE,1);
  int* extent = data->GetExtent();

  for (int y = extent[2]; y <= extent[3]; y++)
  {
    for (int x = extent[0]; x <= extent[1]; x++)
    {
      double* pixel = static_cast<double*>(data->GetScalarPointer(x,y,0));
      pixel[0] = sqrt(pow(x,2.0) + pow(y,2.0));
    }
  }

  vtkSmartPointer<vtkXMLImageDataWriter> writer =
    vtkSmartPointer<vtkXMLImageDataWriter>::New();
  writer->SetFileName("data.vti");
  writer->SetInputData(data);
  writer->Write();
}
}
